### ISSUE TITLE
Is it possible to install a callback on object for unresolved properties?

### ISSUE THREAD

**Issue by swisspol** [cesanta/v7#535]
(sorry more of question than an issue, but I didn't see a `v7` tag in Stack Overflow)

Unless I missed it, it's not possible to do something like this so that a C function would get called back on a JS object whenever JavaScript attempts to access a non-existing property. Such a functionality would allow for more elegant bindings with other languages:

**Before:**
```
var foo = CreateNativeObject("some_class", "some_constructor", arg1, arg2...);
CallNativeObjectMethod(foo, "some_method", arg1, arg2...);
```

**After:**
```
var bar = GetNativeClass("some_class");
var foo = bar.some_constructor(arg1, arg2...) <--- unresolved property callback fires to return JS function wrapping C function
foo.some_method(arg1, arg2...) <--- unresolved property callback fires to return JS function wrapping C function
```

**Comment by swisspol** [comment]
IIRC JavaScript has something like that.

**Comment by swisspol** [comment]
If anyone has any suggestion on how to achieve such elegant bindings, please let me know!

**Comment by swisspol** [comment]
@mmikulicic if you have a few minutes, I'd love your input on this (i.e. getting from "before" to "after" above). Maybe the callback is not the right approach, and a close enough solution can be done today using some JS-foo?

**Comment by mmikulicic** [comment]
AFAIK JS offers a standard way to do it only in ES6. We currently want to complete ES5 support before implementing any ES6 features such as Proxies.

I'd prefer to use a standard API and concepts for that instead of rolling our own, but since it's a C API perhaps it's not necessary to wait for ES6 support.

**Comment by swisspol** [comment]
Got it. In the current `v7.c`, is there an "easy" entry point I could use in the meantime when the code `foo.bar` is executed and the property lookup done? Could I inject in there a custom callback like:
```
v7_val_t LookupMissingProperty(v7_val_t this, const char* name);
```
Presumably that could return a JS function too.

Or by then, it's all byte code and low-level already?

**Comment by mmikulicic** [comment]
hm, let's resume this discussion when we finish our evaluator rewrite. ETA two weeks

Right now the main evaluator is a naive interpreter over a compact AST representation. While it was useful for us to learn javascript it turned out to be a bad choice; we're almost done in transforming it into a bytecode compiler and bytecode evaluator but we still have a few regressions. Stay tuned.

**Comment by swisspol** [comment]
Thanks for the update and good luck with the rewrite!

**Comment by swisspol** [comment]
Feel free to close this issue BTW.

**Comment by kkutsner** [comment]
It looks like there is still no way to set such callback. Any plans to get it implemented?

Here is a use case. I have pretty big array with mega tons of element so it is mapped only partially into memory. I'd like to provide easy access to this array using `[]` operator. Currently, I have to create that array and put all elements there, so it just not possible.

For now, I can expose something like `for_each()` or `get_at()` for my custom array, but it doesn't look nice to use that in the script.

**Comment by mmikulicic** [comment]
I think it's a reasonable feature to ask.

Would the https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Proxy API work for you?

**Comment by dimonomid** [comment]
Resolved by 51b49784cd4ec21de176bfd351eb674aa8d11152: Implemented [Proxy API](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Proxy) (only `get()` and `set()` callbacks are implemented so far).

For JavaScript API, see the link above;
For C API: https://docs.cesanta.com/v7/master/#/c-api/util_public.h/v7_mk_proxy.md/

See usage example of C API in tests: https://github.com/cesanta/v7/blob/51b49784cd4ec21de176bfd351eb674aa8d11152/tests/unit_test.c#L4367
